### ISSUE TITLE
Cannot connect to external SQL Server

### ISSUE THREAD

**Issue by mkilinskidev** [microsoft/navcontainerhelper#2573]
Hi. I am trying to create a new BC container, connected to an external SQL server that is not on the host machine. 

SQL Server IP is 10.0.10.81 and this machine is pingable from the classic bc container, SQLCMD command works as well. But while creating a container with database* parameters there is an error while Importing Encryption Keys

Here is a New-BCContainer prompt
```
New-BcContainer -accept_eula -artifactUrl (Get-BCArtifactUrl -type OnPrem -country w1 -version 20.2 -select Latest) -containerName ext -licenseFile C:\bc\BC20_DEV.flf -updateHosts -shortcuts None -auth UserPassword 
-Credential (New-Object pscredential 'dev', (ConvertTo-SecureString 'Logon2me' -AsPlainText -Force)) 
-databaseServer 10.0.10.81 -databaseInstance ERP 
-databaseCredential (New-Object pscredential 'dev', (ConvertTo-SecureString 'Logon2me' -AsPlainText -Force))
```

And here is a full prompt output
```
BcContainerHelper is version 3.0.11
BcContainerHelper is running as administrator
Hyper-V is Disabled
UsePsSession is True
Host is Microsoft Windows Server 2022 Standard - ltsc2022
Docker Client Version is 20.10.9
Docker Server Version is 20.10.9
Removing Session ext
Removing container ext
Removing entries from hosts
Removing ext from container hosts file
Removing ext-* from container hosts file
Removing Desktop shortcuts
Removing C:\ProgramData\BcContainerHelper\Extensions\ext
Fetching all docker images
Fetching all docker volumes
Using image mcr.microsoft.com/businesscentral:10.0.20348.768
Creating Container ext
Style: onprem
Multitenant: No
Version: 20.2.41144.41423
Platform: 20.0.41086.41354
Generic Tag: 1.0.2.10
Container OS Version: 10.0.20348.768 (ltsc2022)
Host OS Version: 10.0.20348.803 (ltsc2022)
Using process isolation
Using locale en-US
Disabling the standard eventlog dump to container log every 2 seconds (use -dumpEventLog to enable)
Using license file C:\bc\BC20_DEV.flf
Files in C:\ProgramData\BcContainerHelper\Extensions\ext\my:
- AdditionalOutput.ps1
- license.flf
- MainLoop.ps1
- SetupVariables.ps1
- updatehosts.ps1
Creating container ext from image mcr.microsoft.com/businesscentral:10.0.20348.768
98301d1aa96e9d0ecc2d37101bf26d04c68ba2e11b1300ee49ec5cb711b4a0ea
Waiting for container ext to be ready
Using artifactUrl https://bcartifacts.azureedge.net/onprem/20.2.41144.41423/w1
Using installer from C:\Run\150-new
Installing Business Central
Installing from artifacts
Starting Local SQL Server
Starting Internet Information Server
Copying Service Tier Files
c:\dl\onprem\20.2.41144.41423\platform\ServiceTier\Program Files
c:\dl\onprem\20.2.41144.41423\platform\ServiceTier\System64Folder
Copying PowerShell Scripts
c:\dl\onprem\20.2.41144.41423\platform\WindowsPowerShellScripts\Cloud\NAVAdministration
c:\dl\onprem\20.2.41144.41423\platform\WindowsPowerShellScripts\WebSearch
Copying Web Client Files
c:\dl\onprem\20.2.41144.41423\platform\WebClient\Microsoft Dynamics NAV
Copying Client Files
c:\dl\onprem\20.2.41144.41423\platform\LegacyDlls\program files\Microsoft Dynamics NAV
c:\dl\onprem\20.2.41144.41423\platform\LegacyDlls\program files\Microsoft Dynamics NAV
c:\dl\onprem\20.2.41144.41423\platform\LegacyDlls\systemFolder
Copying ModernDev Files
c:\dl\onprem\20.2.41144.41423\platform
c:\dl\onprem\20.2.41144.41423\platform\ModernDev\program files\Microsoft Dynamics NAV
Copying additional files
Copying ConfigurationPackages
C:\dl\onprem\20.2.41144.41423\platform\ConfigurationPackages
Copying Test Assemblies
C:\dl\onprem\20.2.41144.41423\platform\Test Assemblies
Copying Applications
C:\dl\onprem\20.2.41144.41423\platform\Applications
Copying dependencies
Copying ReportBuilder
Importing PowerShell Modules
Skipping restore of Cronus database
Modifying Business Central Service Tier Config File for Docker
Creating Business Central Service Tier
Installing SIP crypto provider: 'C:\Windows\System32\NavSip.dll'
Installation took 29 seconds
Installation complete
Initializing...
Setting host.containerhelper.internal to 172.29.160.1 in container hosts file
Starting Container
Hostname is ext
PublicDnsName is ext
Using NavUserPassword Authentication
Import Encryption Key
Nie mo┼╝na ustanowi─ç po┼é─ůczenia z serwerem/baz─ů danych programu SQL Server.

Mog┼éo tak si─Ö sta─ç z jednej z nast─Öpuj─ůcych przyczyn:

* Program SQL Server nie zosta┼é uruchomiony.
* Baza danych nie istnieje.
* Kontu programu NAV Server nie udzielono odpowiednich uprawnie┼ä.

Spr├│buj ponownie p├│┼║niej lub skontaktuj si─Ö z administratorem systemu.
at <ScriptBlock>, C:\Run\SetupDatabase.ps1: line 152
at <ScriptBlock>, C:\Run\navstart.ps1: line 138
at <ScriptBlock>, C:\Run\start.ps1: line 377
at <ScriptBlock>, <No file>: line 1Error
Using artifactUrl https://bcartifacts.azureedge.net/onprem/20.2.41144.41423/w1
Using installer from C:\Run\150-new
Installing Business Central
Installing from artifacts
Starting Local SQL Server
Starting Internet Information Server
Copying Service Tier Files
c:\dl\onprem\20.2.41144.41423\platform\ServiceTier\Program Files
c:\dl\onprem\20.2.41144.41423\platform\ServiceTier\System64Folder
Copying PowerShell Scripts
c:\dl\onprem\20.2.41144.41423\platform\WindowsPowerShellScripts\Cloud\NAVAdministration
c:\dl\onprem\20.2.41144.41423\platform\WindowsPowerShellScripts\WebSearch
Copying Web Client Files
c:\dl\onprem\20.2.41144.41423\platform\WebClient\Microsoft Dynamics NAV
Copying Client Files
c:\dl\onprem\20.2.41144.41423\platform\LegacyDlls\program files\Microsoft Dynamics NAV
c:\dl\onprem\20.2.41144.41423\platform\LegacyDlls\program files\Microsoft Dynamics NAV
c:\dl\onprem\20.2.41144.41423\platform\LegacyDlls\systemFolder
Copying ModernDev Files
c:\dl\onprem\20.2.41144.41423\platform
c:\dl\onprem\20.2.41144.41423\platform\ModernDev\program files\Microsoft Dynamics NAV
Copying additional files
Copying ConfigurationPackages
C:\dl\onprem\20.2.41144.41423\platform\ConfigurationPackages
Copying Test Assemblies
C:\dl\onprem\20.2.41144.41423\platform\Test Assemblies
Copying Applications
C:\dl\onprem\20.2.41144.41423\platform\Applications
Copying dependencies
Copying ReportBuilder
Importing PowerShell Modules
Skipping restore of Cronus database
Modifying Business Central Service Tier Config File for Docker
Creating Business Central Service Tier
Installing SIP crypto provider: 'C:\Windows\System32\NavSip.dll'
Installation took 29 seconds
Installation complete
Initializing...
Setting host.containerhelper.internal to 172.29.160.1 in container hosts file
Starting Container
Hostname is ext
PublicDnsName is ext
Using NavUserPassword Authentication
Import Encryption Key
Nie mo┼╝na ustanowi─ç po┼é─ůczenia z serwerem/baz─ů danych programu SQL Server.

Mogło tak się stać z jednej z następujących przyczyn:

* Program SQL Server nie został uruchomiony.
* Baza danych nie istnieje.
* Kontu programu NAV Server nie udzielono odpowiednich uprawnień.

Spróbuj ponownie później lub skontaktuj się z administratorem systemu.
at <ScriptBlock>, C:\Run\SetupDatabase.ps1: line 152
at <ScriptBlock>, C:\Run\navstart.ps1: line 138
at <ScriptBlock>, C:\Run\start.ps1: line 377
at <ScriptBlock>, <No file>: line 1
New-BcContainer Telemetry Correlation Id: 738a6f64-aeab-4db9-a68c-9e8871a23434
Initialization of container ext failed
At C:\Program Files\WindowsPowerShell\Modules\BcContainerHelper\3.0.11\ContainerHandling\Wait-NavContainerReady.ps1:42 char:17
+ ...             throw "Initialization of container $containerName failed" ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (Initialization of container ext failed:String) [], RuntimeException
    + FullyQualifiedErrorId : Initialization of container ext failed
```

Maybe because of using a simple login/password credential in databaseCredential parameter but I don't know how to provide the Windows account here. The SQL Server is in the domain. With the sqlcmd I've tried many combinations:
* DOMAIN\user
* user@domain
and nothing works.

**Comment by freddydk** [comment]
Is the SQL Server setup for SQL Server Authentication?
Windows Authentication won't work from inside the container.

**Comment by mkilinskidev** [comment]
SQL Server is configured for mixed auth.

The SQL Server is empty. There are no databases inside it, I don't know it if is important. When I use -replaceExternalDatabase at the begining there is an error

```
BcContainerHelper version 3.0.11
BcContainerHelper emits usage statistics telemetry to Microsoft
BcContainerHelper is version 3.0.11
BcContainerHelper is running as administrator
Hyper-V is Disabled
UsePsSession is True
Host is Microsoft Windows Server 2022 Standard - ltsc2022
Docker Client Version is 20.10.9
Docker Server Version is 20.10.9
Removing Desktop shortcuts
New-BcContainer Telemetry Correlation Id: 0118675a-6d9c-4fa8-89c7-4c7da4083404
Invoke-SqlCmd : The term 'Invoke-SqlCmd' is not recognized as the name of a cmdlet, function, script file, or operable
program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
At C:\Program Files\WindowsPowerShell\Modules\BcContainerHelper\3.0.11\Bacpac\Remove-BcDatabase.ps1:41 char:16
+     $dbFiles = Invoke-SqlCmd `
+                ~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (Invoke-SqlCmd:String) [], ParentContainsErrorRecordException
    + FullyQualifiedErrorId : CommandNotFoundException
```

**Comment by freddydk** [comment]
That error looks like you don't have the SQLPS powershell module on the host?

And yes, I think there need to be a database in order to import an encryption key.

BTW - I don't understand the error message from the first error.

**Comment by mkilinskidev** [comment]
In general, the error says that it can't connect to sql server. But...

When I installed the SQLPS module the above error (with Remove-BcDatabase) doesn't appear but the new-bccontainer prompt stuck on Using NavUserPassword Authentication. 

```
BcContainerHelper is version 3.0.11
BcContainerHelper is running as administrator
Hyper-V is Disabled
UsePsSession is True
Host is Microsoft Windows Server 2022 Standard - ltsc2022
Docker Client Version is 20.10.9
Docker Server Version is 20.10.9
Removing Desktop shortcuts
Starting Database Restore job from https://bcartifacts.azureedge.net/onprem/20.2.41144.41423/w1
Fetching all docker images
Fetching all docker volumes
Using image mcr.microsoft.com/businesscentral:10.0.20348.768
Creating Container ext
Style: onprem
Multitenant: No
Version: 20.2.41144.41423
Platform: 20.0.41086.41354
Generic Tag: 1.0.2.10
Container OS Version: 10.0.20348.768 (ltsc2022)
Host OS Version: 10.0.20348.803 (ltsc2022)
Using process isolation
Using locale en-US
Disabling the standard eventlog dump to container log every 2 seconds (use -dumpEventLog to enable)
Using license file C:\bc\BC20_DEV.flf
Files in C:\ProgramData\BcContainerHelper\Extensions\ext\my:
- AdditionalOutput.ps1
- license.flf
- MainLoop.ps1
- SetupDatabase.ps1
- SetupVariables.ps1
- updatehosts.ps1
Creating container ext from image mcr.microsoft.com/businesscentral:10.0.20348.768
5929541f45985ccb86c87c9508eb8568ce9e2a2b9ddae53b25b408fdf6d983ad
Waiting for container ext to be ready
Using artifactUrl https://bcartifacts.azureedge.net/onprem/20.2.41144.41423/w1
Using installer from C:\Run\150-new
Installing Business Central
Installing from artifacts
Starting Local SQL Server
Starting Internet Information Server
Copying Service Tier Files
c:\dl\onprem\20.2.41144.41423\platform\ServiceTier\Program Files
c:\dl\onprem\20.2.41144.41423\platform\ServiceTier\System64Folder
Copying PowerShell Scripts
c:\dl\onprem\20.2.41144.41423\platform\WindowsPowerShellScripts\Cloud\NAVAdministration
c:\dl\onprem\20.2.41144.41423\platform\WindowsPowerShellScripts\WebSearch
Copying Web Client Files
c:\dl\onprem\20.2.41144.41423\platform\WebClient\Microsoft Dynamics NAV
Copying Client Files
c:\dl\onprem\20.2.41144.41423\platform\LegacyDlls\program files\Microsoft Dynamics NAV
c:\dl\onprem\20.2.41144.41423\platform\LegacyDlls\program files\Microsoft Dynamics NAV
c:\dl\onprem\20.2.41144.41423\platform\LegacyDlls\systemFolder
Copying ModernDev Files
c:\dl\onprem\20.2.41144.41423\platform
c:\dl\onprem\20.2.41144.41423\platform\ModernDev\program files\Microsoft Dynamics NAV
Copying additional files
Copying ConfigurationPackages
C:\dl\onprem\20.2.41144.41423\platform\ConfigurationPackages
Copying Test Assemblies
C:\dl\onprem\20.2.41144.41423\platform\Test Assemblies
Copying Applications
C:\dl\onprem\20.2.41144.41423\platform\Applications
Copying dependencies
Copying ReportBuilder
Importing PowerShell Modules
Skipping restore of Cronus database
Modifying Business Central Service Tier Config File for Docker
Creating Business Central Service Tier
Installing SIP crypto provider: 'C:\Windows\System32\NavSip.dll'
Installation took 36 seconds
Installation complete
Initializing...
Setting host.containerhelper.internal to 172.31.112.1 in container hosts file
Starting Container
Hostname is ext
PublicDnsName is ext
Using NavUserPassword Authentication
```

**Comment by freddydk** [comment]
It might be waiting for the databases to be restored?

**Comment by mkilinskidev** [comment]
Maybe... But (another but :))

My global administrator created this container without any issues with the same parameters, credentials, etc with the below security details
![image](https://user-images.githubusercontent.com/13047961/176624198-f3655209-e0aa-4db1-b586-75c5436038a6.png)
I am wondering why I can't. 

I am a local administrator on the host but not on the sql server, but that shouldn't be a problem. Strange...

**Comment by YannRobertCargo** [comment]
Hello,
I kinda have the same problem.
I've  been wanting to try this but it also gets stuck on "Using NavUserPassword Authentication" with no error.
Both with a "remote" database server and a local sql server on the host.

**Comment by mkilinskidev** [comment]
> Hello, I kinda have the same problem. I've been wanting to try this but it also gets stuck on "Using NavUserPassword Authentication" with no error. Both with a "remote" database server and a local sql server on the host.

In my case, adding my user to the local administrator's group helped.

**Comment by freddydk** [comment]
The issue here is that the database isn't being restored.
But,... the reason why it is hard for you to see what is going wrong is, that the error isn't displayed.

The next version will display:

Setting host.containerhelper.internal to 172.22.32.1 in container hosts file
Starting Container
Hostname is bcserver
PublicDnsName is bcserver
Using NavUserPassword Authentication
**Waiting for database creation to finish**

to symbolize that it is waiting for the restore.
If an error occurs during restore, you will see the error:

Using NavUserPassword Authentication
Waiting for database creation to finish
**Error restoring databases. Error was The term 'Download-Artifacts' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.**
Import Encryption Key

It will continue to try and use the database (and import encryption key), which will fail if the database doesn't exist.

So - the next version won't change the functionality but will help troubleshooting.

**Comment by freddydk** [comment]
fixed in the latest preview

**Comment by freddydk** [comment]
Shipped in 4.0.3

**Comment by SjoerddeClippelaar** [comment]
@freddydk the changes for this issue do something with Add and Remove of AppDomain (and the code to determine $OnAssemblyResolve). But this is giving errors locally. Do you remember why you did this? Removing it solves my issue importing the dlls that are not in the folder.

[System.AppDomain]::CurrentDomain.add_AssemblyResolve($OnAssemblyResolve)
